### ISSUE TITLE
rc94 discussion (mid-March 2021?)

### ISSUE THREAD

**Issue by kolyshkin** [opencontainers/runc#2790]
After releasing rc93, we are in feature freeze, i.e. only bug fixes¹ are accepted into rc94,
which (hopefully) will be the last rc before the final 1.0 release.

Issues and PRs targeted for rc94 need to be added to [1.0.0-rc94 milestone](https://github.com/opencontainers/runc/milestone/11)

Changes since rc93: https://github.com/opencontainers/runc/compare/v1.0.0-rc93...master

¹ I want to make an exception for CI improvements, as those do not directly affect the code,
but might result in better quality/coverage/etc.

**Comment by cyphar** [comment]
Yeah, i'm fine with CI changes. Feature freeze doesn't include things like documentation and CI. Also depending on Docker's testing we might be able to skip rc94 entirely.

**Comment by JonathonReinhart** [comment]
Am I too late for #2826 to be considered for v1.0.0?

**Comment by h-vetinari** [comment]
Mid-march has arrived - there's still a fair amount of open PRs in the [milestone](https://github.com/opencontainers/runc/milestone/11). Biggest unknown seems to be #2797, which doesn't have a PR yet.

PS. June 4th is the 5th birthday of [`1.0.0-rc1`](https://github.com/opencontainers/runc/releases/tag/v1.0.0-rc1), would be fun if 1.0.0 could/would be released on that date 🙃

**Comment by cyphar** [comment]
I should have a PR for #2797 ready next week or so -- I was on vacation hence the lack of work on it. :wink:

**Comment by liftM** [comment]
Is there an ETA for the rc94 release? I'm particularly interested in https://github.com/opencontainers/runc/issues/2801#issuecomment-776190119 since I'm on Arch and Docker is just straight up broken for me until this patch is released. If the release is soon, I'll just wait for it, but if it's far off, I'm going to look at messing with my kernel parameters instead.

**Comment by kolyshkin** [comment]
Just talked to @cyphar  -- it seems it makes sense to release rc94 before the cgroup v2 device changes he is working on are ready, as we already have a decent amount of fixes; that includes a fix for regression in rc93 (issues #2865, #2828, presumably fixed by PR #2871).

Let's try to merge the remaining PRs with the rc94 milestone and prepare a release.

**Comment by cyphar** [comment]
Yup, I can cook up a release with those fixes once they're merged.

**Comment by AkihiroSuda** [comment]
SGTM

**Comment by haircommander** [comment]
any opposition to adding https://github.com/opencontainers/runc/pull/2894 (fixes a regression between rc92 and rc93)

**Comment by mrunalp** [comment]
SGTM

**Comment by cyphar** [comment]
Only big thing remaining is #2896.

**Comment by kolyshkin** [comment]
I'd like to have https://github.com/opencontainers/runc/pull/2897 included as this fixes a real bug (albeit not a regression) and I'd rather have the fix being tested in rc94.

**Comment by h-vetinari** [comment]
It seems the usual "just one more PR" also applies to this release, but the end seems to be in sight now - how about pushing this over the finish line?

**Comment by tao12345666333** [comment]
SGTM 
Looking at milestone [1.0.0-rc94](https://github.com/opencontainers/runc/milestone/11) , we will find that only #2906 is left now. Will we release rc94 after #2906 is completed?

**Comment by AkihiroSuda** [comment]
We want to fix the performance regression in rc93 before releasing rc94: https://github.com/opencontainers/runc/pull/2921

**Comment by cyphar** [comment]
Everything is merged. I will prepare a release PR on top of d279ebd97d8832020e2c6f50cc3a11d0499a4690.

**Comment by kolyshkin** [comment]
For the changelog, this is what I have based only on `impact/changelog` labels set:

Notable changes:
* cgroupv1: kernel memory limits are now ignored (#2840)
* libcontainer/cgroups: cgroup managers' `Set` now accept `configs.Resources` rather than `configs.Cgroups` (#2906)
* libcontainer/cgroups/systemd: reconnect and retry in case dbus connection is closed (after dbus restart) (#2923)

Bugfixes:
* seccomp: fix 32-bit compilation errors (regression in rc93, #2783)
* cgroupv2: blkio weight value conversion fix (#2786)
* runc init: fix a hang caused by deadlock in seccomp/ebpf loading code (regression in rc93, #2871)
* runc start: fix "chdir to cwd: permission denied"  for some setups (regression in rc93, #2894)
* s390: fix broken terminal (regression in rc93, #2898)

Improvements:
* runc start/exec: better diagnostics when container limits are too low (#2812)
* cgroupv2: multiple GetStats improvements (#2816, #2873)
* cgroupv2: fallback to setting io.weight if io.bfq.weight is not available (#2820)
* capabilities: WARN, not ERROR, for unknown / unavailable capabilities (#2854)

**Comment by cpuguy83** [comment]
Callout for https://github.com/opencontainers/runc/issues/2928 ?

**Comment by cyphar** [comment]
It would be nice if we'd have a proper `CHANGELOG.md` so I didn't have to write one from scratch each time.

**Comment by h-vetinari** [comment]
> 
> 
> It would be nice if we'd have a proper `CHANGELOG.md` so I didn't have to write one from scratch each time.

Ummmmm... #2368. I offered to do this (including backfilling for a few releases using a to-be-agreed [format](https://github.com/opencontainers/runc/pull/2368#issuecomment-622348906)), but the appetite was very low.

**Comment by cyphar** [comment]
Yeah I remember the issue, the problem was that back-filling the changelog with prettified git log messages isn't quite ideal. Back-filling it with the actual release notes is probably more useful to be honest. But I don't even think we need to backfill it, we could just start a new changelog from scratch with the next release and add stuff with each PR.

**Comment by h-vetinari** [comment]
> [...] back-filling the changelog with prettified git log messages isn't quite ideal. Back-filling it with the actual release notes is probably more useful to be honest.

That was the only starting point; could have been brought into better shape (e.g. the format in the linked comment) during the PR, and including previous release notes would be trivial.

**Comment by kolyshkin** [comment]
> It would be nice if we'd have a proper CHANGELOG.md so I didn't have to write one from scratch each time.

My proposal is to
1. Have impact/changelog label for every PR that is worth mentioning (bug fix, notable improvement, new feature, potentially breaking change etc).
2. Have a "Changelog entry" field in such PRs.
3. collect those entries before the release.

This is sort of what I did. Was not perfect but AFAICS it works.

**Comment by tao12345666333** [comment]
I agree with @kolyshkin ‘s proposal. We can reach an agreement during the PR merger without having to count again afterwards.

**Comment by cyphar** [comment]
@h-vetinari The reason I didn't put that note is because I knew we would be doing another release today to fix the CVE. There are only two PRs in the 1.0.0 milestone, we will do a release once they are merged (barring any other critical fixes).

**Comment by h-vetinari** [comment]
> The reason I didn't put that note is because I knew we would be doing another release today to fix the CVE

Makes complete sense, but I couldn't know that 😅 

Very happy to hear it, amazing to see this (leg of the) journey come to an end! 😊
